File: src/main/java/io/scriptor/loader/ResourceClassLoader.java
package io.scriptor.loader;

import io.scriptor.log.Log;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import static io.scriptor.loader.IPackageEntry.*;

public class ResourceClassLoader extends ClassLoader {

    public ResourceClassLoader(final @Nullable ClassLoader parentLoader) {
        super(parentLoader);
    }

    public @NotNull IterableEntry getClassPackage(final @NotNull String packageName)
            throws IOException, ClassNotFoundException {
        final var fileName = packageName.replace('.', File.separatorChar);
        try (final var packageStream = getResourceAsStream(fileName)) {
            if (packageStream == null) {
                return new IterableEntry();
            }

            final List<IPackageEntry> entries = new ArrayList<>();

            final var reader = new BufferedReader(new InputStreamReader(packageStream));
            for (String line; (line = reader.readLine()) != null; ) {
                final var lineNoClass = line.replaceAll("\\.class", "");
                final var entryName = packageName.isEmpty()
                                      ? lineNoClass
                                      : "%s.%s".formatted(packageName, lineNoClass);

                if (!line.endsWith(".class")) {
                    entries.add(new FutureEntry(this, entryName));
                    continue;
                }

                entries.add(new ClassEntry(loadClass(entryName)));
            }

            return new IterableEntry(entries);
        }
    }

    @Override
    protected @NotNull Class<?> findClass(final @NotNull String className) throws ClassNotFoundException {
        Log.info("query class %s", className);
        final var fileName = className.replace('.', File.separatorChar) + ".class";
        try (final var classStream = getResourceAsStream(fileName)) {
            final var classBytes = Objects.requireNonNull(classStream).readAllBytes();
            return defineClass(className, classBytes, 0, classBytes.length);
        } catch (final IOException e) {
            throw new ClassNotFoundException(e.getMessage());
        }
    }
}

File: src/main/java/io/scriptor/loader/ResourceClassLoader.kt
package io.scriptor.loader

import io.scriptor.loader.IPackageEntry.FutureEntry
import io.scriptor.loader.IPackageEntry.IterableEntry
import io.scriptor.log.info
import java.io.BufferedReader
import java.io.File
import java.io.IOException
import java.io.InputStreamReader

open class ResourceClassLoader(parentLoader: ClassLoader?) : ClassLoader(parentLoader) {

    open fun getClassPackage(packageName: String): IterableEntry {
        val filename = packageName.replace('.', File.separatorChar)

        getResourceAsStream(filename).use { packageStream ->
            if (packageStream == null) {
                return IterableEntry()
            }

            val entries: MutableList<IPackageEntry> = ArrayList()

            val reader = BufferedReader(InputStreamReader(packageStream))
            var line: String?
            while ((reader.readLine().also { line = it }) != null) {
                val lineNoClass = line!!.replace("\\.class".toRegex(), "")
                val entryName =
                    if (packageName.isEmpty()) lineNoClass
                    else "%s.%s".format(packageName, lineNoClass)

                if (!line.endsWith(".class")) {
                    entries.add(FutureEntry(this, entryName))
                    continue
                }

                entries.add(IPackageEntry.ClassEntry(loadClass(entryName)))
            }
            return IterableEntry(entries)
        }
    }

    override fun findClass(className: String): Class<*> {
        info("query class %s", className)
        val fileName = className.replace('.', File.separatorChar) + ".class"
        try {
            getResourceAsStream(fileName).use { classStream ->
                val classBytes = classStream!!.readAllBytes()
                return defineClass(className, classBytes, 0, classBytes.size)
            }
        } catch (e: IOException) {
            throw ClassNotFoundException(e.message)
        }
    }
}
